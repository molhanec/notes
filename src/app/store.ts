import { Action, configureStore, getDefaultMiddleware, ThunkAction } from '@reduxjs/toolkit'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { createBrowserHistory } from 'history'

import counterReducer from '../features/counter/counterSlice'
import notesReducer from '../features/notes/notesSlice'

export const history = createBrowserHistory()

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    router: connectRouter(history) as any,
    notes: notesReducer,
  },
  middleware: [...getDefaultMiddleware(), routerMiddleware(history)],
})

export type RootState = ReturnType<typeof store.getState>
export type AppThunk = ThunkAction<void, RootState, unknown, Action<string>>
