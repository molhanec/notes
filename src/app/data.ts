import { AppThunk } from "./store"
import { loadNotesAsync } from "../features/noteList/notesSlice"
import { setProgress, setSuccess, setError } from "../features/info/infoSlice"

const rootDataUrl =
  "https://private-anon-3ab0da907c-note10.apiary-mock.com/notes"

const checkResponseStatus = (response: Response) => {
  if (response.ok) {
    return response
  } else {
    return Promise.reject(
      `Error fetching data: ${response.status} ${response.statusText}`
    )
  }
}

export const fetchJson = <T>(path: string) =>
  fetch(rootDataUrl + path)
    .then(checkResponseStatus)
    .then(response => response!.json() as Promise<T>)

export const addNoteAsync = (title: string): AppThunk => async dispatch => {
  dispatch(setProgress())
  try {
    fetch(rootDataUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title,
      }),
    }).then(checkResponseStatus)

    dispatch(setSuccess("New note added"))
    dispatch(loadNotesAsync())
  } catch {
    dispatch(setError("Cannot add new note"))
  }
}
