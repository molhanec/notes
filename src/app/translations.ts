import { useSelector } from "react-redux"
import { RootState } from "./store"

export type LanguageCodes = "cz" | "en"
export const defaultLanguage = "cz"

interface Phrases {
  appTitle: string
  noteLoadFailure: string
  notesLoadFailure: string

  noteCreateSuccess: string
  noteCreateFailure: string
  noteUpdateSuccess: string
  noteUpdateFailure: string
  noteDeleteSuccess: string
  noteDeleteFailure: string

  formTitle: string
  formSave: string
  formDelete: string

  editFormHeader: string
  newFormHeader: string
}

export type PhraseKeys = keyof Phrases

const enPhrases: Phrases = {
  appTitle: "Notes",
  noteLoadFailure: "Cannot load note",
  notesLoadFailure: "Cannot load notes",

  noteCreateSuccess: "New note added",
  noteCreateFailure: "Cannot add new note",
  noteUpdateSuccess: "Note updated",
  noteUpdateFailure: "Cannot update note",
  noteDeleteSuccess: "Note deleted",
  noteDeleteFailure: "Cannot delete note",

  formTitle: "Title",
  formSave: "Save",
  formDelete: "Delete",

  editFormHeader: "Note details",
  newFormHeader: "Create new note",
}

const czPhrases: Phrases = {
  appTitle: "Poznámky",
  noteLoadFailure: "Nelze načíst poznámku",
  notesLoadFailure: "Nelze načíst poznámky",

  noteCreateSuccess: "Přidána poznámka",
  noteCreateFailure: "Nelze přidat poznámku",
  noteUpdateSuccess: "Poznámka upravena",
  noteUpdateFailure: "Nelze upravit poznámku",
  noteDeleteSuccess: "Poznámka smazána",
  noteDeleteFailure: "Nelze smazat poznámku",

  formTitle: "Nadpis",
  formSave: "Uložit",
  formDelete: "Smazat",

  editFormHeader: "Detaily poznámky",
  newFormHeader: "Vytvořit novou poznámku",
}

export const getPhrase = (language: "cz" | "en", key: PhraseKeys): string => {
  switch (language) {
    case "cz":
      return czPhrases[key]
    case "en":
      return enPhrases[key]
  }
}

export const useTranslate = () => {
  const language = useSelector((state: RootState) => state.language.code)
  return (key: PhraseKeys) => getPhrase(language, key)
}
