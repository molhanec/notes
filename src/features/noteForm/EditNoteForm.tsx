import React from "react"
import { connect, ResolveThunks } from "react-redux"
import { updateNoteAsync } from "../../app/data"
import { NoteForm } from "./NoteForm"
import { RootState } from "../../app/store"

type Props = ReturnType<typeof mapStateToProps> &
  ResolveThunks<typeof mapDispatchToProps>

export const EditNoteForm = ({
  noteDetails,
  progress,
  updateNoteAsync,
}: Props) => {
  const handleSubmit = (title: string): string => {
    if (!noteDetails.id) return title
    updateNoteAsync(noteDetails.id, title)
    return title
  }

  return (
    <NoteForm
      disabled={!noteDetails || progress}
      onSubmit={handleSubmit}
      initialTitle={noteDetails.title || ""}
    />
  )
}

const mapStateToProps = (state: RootState) => ({
  noteDetails: state.noteDetails,
  progress: state.info.progress,
})

const mapDispatchToProps = {
  updateNoteAsync,
}

const ConnectedNewNoteForm = connect(
  mapStateToProps,
  mapDispatchToProps
)(EditNoteForm)
export default ConnectedNewNoteForm
