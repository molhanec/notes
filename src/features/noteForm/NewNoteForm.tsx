import React from "react"
import { connect, ResolveThunks } from "react-redux"
import { addNoteAsync } from "../../app/data"
import { NoteForm } from "./NoteForm"
import { RootState } from "../../app/store"
import { useTranslate } from "./../../app/translations"

type Props = ReturnType<typeof mapStateToProps> &
  ResolveThunks<typeof mapDispatchToProps>

export const NewNoteForm = ({ addNoteAsync, progress }: Props) => {
  const t = useTranslate()

  const handleSubmit = (title: string) => addNoteAsync(title)

  return (
    <NoteForm
      formHeader={t("newFormHeader")}
      onSubmit={handleSubmit}
      disabled={progress}
    />
  )
}

const mapStateToProps = (state: RootState) => ({
  progress: state.info.progress,
})

const mapDispatchToProps = {
  addNoteAsync,
}

const ConnectedNewNoteForm = connect(
  mapStateToProps,
  mapDispatchToProps
)(NewNoteForm)
export default ConnectedNewNoteForm
