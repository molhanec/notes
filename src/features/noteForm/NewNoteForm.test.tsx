import React from "react"
import { render, fireEvent } from "@testing-library/react"
import { Provider } from "react-redux"
import { store } from "../../app/store"
import { EditNoteForm } from "./EditNoteForm"
import { getPhrase } from "../../app/translations"
import { NewNoteForm } from "./NewNoteForm"

describe("<NewNoteForm>", () => {
  test("matches snapshot", () => {
    const { container } = render(
      <Provider store={store}>
        <NewNoteForm progress={false} addNoteAsync={jest.fn()} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test("calls addNoteAsync for non-empty title when Save button clicked", () => {
    const addNoteAsync = jest.fn()
    const { getByText, getByDisplayValue } = render(
      <Provider store={store}>
        <NewNoteForm progress={false} addNoteAsync={addNoteAsync} />
      </Provider>
    )

    const input = getByDisplayValue("")
    fireEvent.change(input, { target: { value: "X" } })

    const saveButton = getByText(getPhrase("cz", "formSave"))
    fireEvent.click(saveButton)

    expect(addNoteAsync).toBeCalled()
  })

  test("does not call addNoteAsync for empty title when Save button clicked", () => {
    const addNoteAsync = jest.fn()
    const { getByText } = render(
      <Provider store={store}>
        <NewNoteForm progress={false} addNoteAsync={addNoteAsync} />
      </Provider>
    )

    const saveButton = getByText(getPhrase("cz", "formSave"))
    fireEvent.click(saveButton)
    expect(addNoteAsync).not.toBeCalled()
  })
})
