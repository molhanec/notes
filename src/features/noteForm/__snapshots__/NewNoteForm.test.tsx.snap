// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewNoteForm> matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Vytvořit novou poznámku
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <label
          class=""
          for="title"
        >
          Nadpis
        </label>
        <input
          class="form-control"
          id="title"
          type="text"
          value=""
        />
        <button
          class="mt-3 btn btn-primary disabled"
          disabled=""
        >
          Uložit
        </button>
      </form>
    </div>
  </div>
</div>
`;
